File: src/com/developerphil/adbidea/ui/DeviceChooserDialog.java
package com.developerphil.adbidea.ui;

import com.android.ddmlib.IDevice;
import com.intellij.ide.util.PropertiesComponent;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.openapi.util.Disposer;
import org.jetbrains.android.facet.AndroidFacet;
import org.jetbrains.android.run.DeviceChooser;
import org.jetbrains.android.run.DeviceChooserListener;
import org.jetbrains.android.util.AndroidBundle;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

public class DeviceChooserDialog extends DialogWrapper {
    private final Project myProject;
    private final DeviceChooser myDeviceChooser;

    private JPanel myPanel;
    private JPanel myDeviceChooserWrapper;

    @NonNls
    private static final String SELECTED_SERIALS_PROPERTY = DeviceChooserDialog.class.getCanonicalName() + "-SELECTED_DEVICES";

    public DeviceChooserDialog(@NotNull final AndroidFacet facet, boolean multipleSelection) {
        super(facet.getModule().getProject(), true);
        setTitle(AndroidBundle.message("choose.device.dialog.title"));

        myProject = facet.getModule().getProject();
        final PropertiesComponent properties = PropertiesComponent.getInstance(myProject);

        final String[] selectedSerials;
        final String serialsStr = properties.getValue(SELECTED_SERIALS_PROPERTY);
        if (serialsStr != null) {
            selectedSerials = serialsStr.split(" ");
        } else {
            selectedSerials = null;
        }

        getOKAction().setEnabled(false);

        myDeviceChooser = new DeviceChooser(multipleSelection, getOKAction(), facet, null);
        Disposer.register(myDisposable, myDeviceChooser);
        myDeviceChooser.addListener(new DeviceChooserListener() {
            @Override
            public void selectedDevicesChanged() {
                updateOkButton();
            }
        });

        myDeviceChooserWrapper.add(myDeviceChooser.getPanel());

        init();

        myDeviceChooser.init(selectedSerials);

        updateEnabled();
    }

    private void updateOkButton() {
        getOKAction().setEnabled(getSelectedDevices().length > 0);
    }

    private void updateEnabled() {
        updateOkButton();
    }

    @Override
    public JComponent getPreferredFocusedComponent() {
        return myDeviceChooser.getDeviceTable();
    }

    @Override
    protected void doOKAction() {
        myDeviceChooser.finish();

        final PropertiesComponent properties = PropertiesComponent.getInstance(myProject);
        properties.setValue(SELECTED_SERIALS_PROPERTY, toString(myDeviceChooser.getSelectedDevices()));

        super.doOKAction();
    }

    @Override
    protected String getDimensionServiceKey() {
        return getClass().getCanonicalName();
    }

    @Override
    protected JComponent createCenterPanel() {
        return myPanel;
    }

    @NotNull
    public IDevice[] getSelectedDevices() {
        return myDeviceChooser.getSelectedDevices();
    }

    @NotNull
    public static String toString(@NotNull IDevice[] devices) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0, n = devices.length; i < n; i++) {
            builder.append(devices[i].getSerialNumber());
            if (i < n - 1) {
                builder.append(' ');
            }
        }
        return builder.toString();
    }

}

File: src/main/kotlin/com/developerphil/adbidea/ui/DeviceChooserDialog.kt
package com.developerphil.adbidea.ui

import com.android.ddmlib.IDevice
import com.developerphil.adbidea.ObjectGraph
import com.developerphil.adbidea.preference.ProjectPreferences
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.util.Disposer
import org.jetbrains.android.facet.AndroidFacet
import org.jetbrains.android.util.AndroidBundle
import org.joor.Reflect
import java.awt.BorderLayout
import javax.swing.JCheckBox
import javax.swing.JComponent
import javax.swing.JPanel

/**
 * Programmatic UI for the device chooser dialog. Replaces the previous UI Designer-based
 * approach to avoid initialization order issues across IDE versions.
 */
class DeviceChooserDialog(facet: AndroidFacet) : DialogWrapper(facet.module.project, true) {

    private val myProject: Project = facet.module.project
    private val projectPreferences: ProjectPreferences =
        myProject.getService(ObjectGraph::class.java).projectPreferences

    private val rootPanel = JPanel(BorderLayout())
    private val deviceChooserWrapper = JPanel(BorderLayout())
    private val useSameDeviceCheckBox = JCheckBox("Use same device(s) for future commands")

    private val myDeviceChooser: MyDeviceChooser

    val selectedDevices: Array<IDevice>
        get() = myDeviceChooser.selectedDevices

    init {
        title = AndroidBundle.message("choose.device.dialog.title")
        okAction.isEnabled = false

        myDeviceChooser = MyDeviceChooser(true, okAction, facet, null)
        Disposer.register(myDisposable, myDeviceChooser)
        myDeviceChooser.addListener(object : DeviceChooserListener {
            override fun selectedDevicesChanged() {
                updateOkButton()
            }
        })

        // Build static UI
        rootPanel.add(deviceChooserWrapper, BorderLayout.CENTER)
        rootPanel.add(useSameDeviceCheckBox, BorderLayout.SOUTH)

        // Initialize DialogWrapper and dynamic content
        init()
        deviceChooserWrapper.add(myDeviceChooser.panel, BorderLayout.CENTER)
        myDeviceChooser.init(projectPreferences.getSelectedDeviceSerials())
        updateOkButton()
    }

    private fun persistSelectedSerialsToPreferences() {
        projectPreferences.saveSelectedDeviceSerials(
            myDeviceChooser.selectedDevices.map { it.serialNumber }.toList()
        )
    }

    private fun updateOkButton() {
        okAction.isEnabled = selectedDevices.isNotEmpty()
    }

    override fun getPreferredFocusedComponent(): JComponent? {
        return try {
            myDeviceChooser.preferredFocusComponent
        } catch (e: NoSuchMethodError) { // preview versions fallback
            Reflect.on(myDeviceChooser).call("getDeviceTable").get<JComponent>()
        }
    }

    override fun doOKAction() {
        myDeviceChooser.finish()
        persistSelectedSerialsToPreferences()
        super.doOKAction()
    }

    override fun getDimensionServiceKey() = javaClass.canonicalName

    override fun createCenterPanel(): JComponent = rootPanel

    fun useSameDevices() = useSameDeviceCheckBox.isSelected
}